src/probability/kernel/disintegration.lean

```diff
@@ -0,0 +1,1214 @@
+    obtain ⟨y, hxy, hyr⟩ := exists_rat_btwn r.prop,
+    refine cinfi_set_le hf (hxy.trans _),
+    exact_mod_cast hyr, },
+  { refine le_cinfi (λ q, _),
+    have hq := q.prop,
+    rw mem_Ioi at hq,
+    obtain ⟨y, hxy, hyq⟩ := exists_rat_btwn hq,
+    refine (cinfi_le _ _).trans _,
+    { exact ⟨y, hxy⟩, },
+    { refine ⟨hf.some, λ z, _⟩,
+      rintros ⟨u, rfl⟩,
+      suffices hfu : f u ∈ f '' Ioi x, from hf.some_spec hfu,
+      exact ⟨u, u.prop, rfl⟩, },
+    { refine hf_mono (le_trans _ hyq.le),
+      norm_cast, }, },
+end
+
+lemma lintegral_sub' {m : measurable_space α} {μ : measure α} {f g : α → ℝ≥0∞}
+  (hg : ae_measurable g μ) (hg_fin : ∫⁻ a, g a ∂μ ≠ ∞) (h_le : g ≤ᵐ[μ] f) :
+  ∫⁻ a, f a - g a ∂μ = ∫⁻ a, f a ∂μ - ∫⁻ a, g a ∂μ :=
+begin
+  refine ennreal.eq_sub_of_add_eq hg_fin _,
+  rw [← lintegral_add_right' _ hg],
+  exact lintegral_congr_ae (h_le.mono $ λ x hx, tsub_add_cancel_of_le hx)
+end
+
+lemma ennreal.tendsto_at_top_at_bot [nonempty ι] [semilattice_sup ι]
+  {f : ι → ℝ≥0∞} (h : tendsto f at_top at_bot) :
+  tendsto f at_top (𝓝 0) :=
+begin
+  rw tendsto_at_bot at h,
+  specialize h 0,
+  rw eventually_at_top at h,
+  obtain ⟨i, hi⟩ := h,
+  rw ennreal.tendsto_at_top_zero,
+  exact λ ε hε, ⟨i, λ n hn, (hi n hn).trans (zero_le _)⟩,
+end
+
+lemma tendsto_of_antitone {ι α : Type*} [preorder ι] [topological_space α]
+  [conditionally_complete_linear_order α] [order_topology α] {f : ι → α} (h_mono : antitone f) :
+  tendsto f at_top at_bot ∨ (∃ l, tendsto f at_top (𝓝 l)) :=
+@tendsto_of_monotone ι αᵒᵈ _ _ _ _ _ h_mono
+
+lemma tendsto_at_top_at_bot_iff_of_antitone [nonempty α] [semilattice_sup α] [preorder β]
+  {f : α → β} (hf : antitone f) :
+  tendsto f at_top at_bot ↔ ∀ b : β, ∃ a : α, f a ≤ b :=
+@tendsto_at_top_at_top_iff_of_monotone α βᵒᵈ _ _ _ _ hf
+
+lemma image_fst_prod (s : set α) (t : set β) (ht : t.nonempty) :
+  prod.fst '' s ×ˢ t = s :=
+begin
+  ext1 x,
+  simp only [mem_image, mem_prod, prod.exists, exists_and_distrib_right,
+    exists_and_distrib_left, exists_eq_right, and_iff_left_iff_imp],
+  exact λ _, ht,
+end
+
+lemma to_real_infi (f : α → ℝ≥0∞) (hf : ∀ a, f a ≠ ∞) :
+  (⨅ i, f i).to_real = ⨅ i, (f i).to_real :=
+begin
+  casesI is_empty_or_nonempty α,
+  { -- todo: real.cinfi_empty should be a simp lemma
+    simp only [with_top.cinfi_empty, ennreal.top_to_real, real.cinfi_empty], },
+  have h_ne_top : (⨅ i, f i) ≠ ∞,
+  { refine ne_of_lt (lt_of_le_of_lt _ (hf h.some).lt_top),
+    exact infi_le _ _, },
+  refine le_antisymm _ _,
+  { refine le_cinfi (λ a, (ennreal.to_real_le_to_real h_ne_top (hf a)).mpr _),
+    exact infi_le _ _, },
+  { rw ← @ennreal.to_real_of_real (⨅ i, (f i).to_real),
+    swap, { exact le_cinfi (λ i, ennreal.to_real_nonneg), },
+    rw ennreal.to_real_le_to_real ennreal.of_real_ne_top h_ne_top,
+    refine le_infi (λ i, ennreal.of_real_le_of_le_to_real _),
+    refine cinfi_le ⟨0, λ i hi, _⟩ i,
+    obtain ⟨j, rfl⟩ := mem_range.mpr hi,
+    exact ennreal.to_real_nonneg, },
+end
```

Comment:
```suggestion
begin
  casesI is_empty_or_nonempty α,
  { -- todo: real.cinfi_empty should be a simp lemma
    simp only [with_top.cinfi_empty, ennreal.top_to_real, real.cinfi_empty], },
  lift f to α → ℝ≥0 using hf,
  simp_rw [← with_top.coe_infi, ennreal.coe_to_real, nnreal.coe_infi],
end
```

Reply:
Thanks! That's a useful tactic. I forgot it existed.